product_default_image/models/product_template.py

```diff
@@ -0,0 +1,100 @@
+        '(False prevents change)',
+    )
+
+    @api.model
+    def default_get(self, vals):
+        vals = super(ProductTemplate, self).default_get(vals)
+
+        if not vals.get('categ_id'):
+            return vals
+
+        company = self.env.user.company_id
+        category_mod = self.env['product.category']
+        target = company.product_image_target
+
+        if target == 'category':
+            categ = category_mod.browse(vals['categ_id'])
+            vals['categ_change_image'] = True
+            vals['image'] = categ.image
+            vals['image_medium'] = categ.image_medium
+            vals['image_small'] = categ.image_small
+
+        elif target == 'global':
+            vals['image'] = company.product_default_global_image
+            tools.image_resize_images(vals)
+
+        return vals
+
+    @api.multi
+    @api.onchange('categ_id')
+    def _onchange_categ_id(self):
+        """ Changes image to new categ_id's image.
+
+        This method checks if a custom image has been uploaded. If not,
+        the image is changed to the new categ_id's image.
+
+        Cases where it is assumed a custom image has NOT been uploaded:
+            * image, image_medium, and image_small are all empty
+            * image matches any category's image field
+            * image_medium matches any category's image_medium field
+            * image_small matches any category's image_small field
+
+        image, image_medium, and image_small fields are all compared against
+        categories because, for example, in certain cases the image field
+        is None, but the image_medium field has data.
+        """
+        for record in self:
+
+            if not self.env.user.company_id.product_image_target == 'category':
+                continue
+
+            if not record.categ_change_image:
+                continue
+
+            categ_mod = self.env['product.category']
+            categories = categ_mod.search([])
+
+            image_fields = {
+                'image': record.image,
+                'image_medium': record.image_medium,
+                'image_small': record.image_small,
+            }
+
+            matched_categs = []
+            empty_image_field = []
+
+            for key, value in image_fields.iteritems():
+                if matched_categs:
+                    break
+
+                if not value:
+                    empty_image_field.append(True)
+                    continue
+
+                matched_categs += filter(
+                    lambda s: getattr(s, key, False) == value, categories
+                )
+
+            if matched_categs or len(empty_image_field) == 3:
+                record.image = record.categ_id.image
+                record.image_medium = record.categ_id.image_medium
+                record.image_small = record.categ_id.image_small
+
+    @api.model
+    def create(self, vals):
+        vals['categ_change_image'] = False
```

Comment:
This still should be False, as it's set to True in default_get (although I should probs just set `default=True` right on the field.

This just prevents the product's image from being changed when the user changes the categ_id after the product has been created.